cmd/vela-worker/register.go

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2020 Target Brands, Inc. All rights reserved.
+//
+// Use of this source code is governed by the LICENSE file in this repository.
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-vela/types/library"
+	"github.com/sirupsen/logrus"
+)
+
+// register is a helper function to register
+// the worker in the database, updating the item
+// if the worker already exists
+func (w *Worker) register(config *library.Worker) error {
+	// check to see if the worker already exists in the database
+	_, resp, err := w.VelaClient.Worker.Get(config.GetHostname())
+	if err != nil {
+		// check to see if the response was nill
+		if resp == nil {
+			return fmt.Errorf("unable to retrieve worker %s from the server: %v", config.GetHostname(), err)
+		}
+		// check to see if the worker was not found and if we need to add it
+		if resp.StatusCode == http.StatusNotFound {
+			logrus.Infof("registering worker %s with the server", config.GetHostname())
+			_, _, err := w.VelaClient.Worker.Add(config)
+			if err != nil {
+				// log the error instead of returning so the operation doesn't block worker deployment
+				return fmt.Errorf("unable to register worker %s with the server: %v", config.GetHostname(), err)
+			}
+		} else {
+			return fmt.Errorf("unable to retrieve worker %s from the server: %v", config.GetHostname(), err)
+		}
```

Comment:
Do we want to `return nil` after we add the worker?

If I'm reading the code correctly, we attempt to get the worker to see if it exists.

If it doesn't exist, we check to see if the status code returned was a `404`(not found):
  * if status code == `404`, then we attempt to add the worker to Vela
  * if status code != `404`, then we return an error

I think this all sounds good, but the way it is written, I believe that after we successfully add the worker, then we'll exit out of the `if` block and continue processing the rest of this function.

The end result would be after we make the call to add the worker, we would then continue to the logic that updates the worker at the end of this function meaning that we make back to back API calls.

I know this is a book so I think I'll show what I'm suggesting:

```suggestion
		// check to see if the worker was not found and if we need to add it
		if resp.StatusCode == http.StatusNotFound {
			logrus.Infof("registering worker %s with the server", config.GetHostname())
			_, _, err := w.VelaClient.Worker.Add(config)
			if err != nil {
				// log the error instead of returning so the operation doesn't block worker deployment
				return fmt.Errorf("unable to register worker %s with the server: %v", config.GetHostname(), err)
			}
			
			// successfully added the worker so return nil
			return nil
		}
		
		return fmt.Errorf("unable to retrieve worker %s from the server: %v", config.GetHostname(), err)
```

Reply:
Good catch! You looks to be correct 🤕